utils/create_dataset.py

```diff
@@ -104,3 +114,155 @@ def __getitem__(self, index):
+        Args:
+            query:
+
+        Returns: Torch tensor patch.
+
+        """
+        bounds = (query.minx, query.miny, query.maxx, query.maxy)
+        out_width = round((query.maxx - query.minx) / self.res)
+        out_height = round((query.maxy - query.miny) / self.res)
+        out_shape = (self.vrt_ds.count, out_height, out_width)
+
+        dest = self.vrt_ds.read(
+            out_shape=out_shape, window=from_bounds(*bounds, self.vrt_ds.transform)
+        )
+
+        if dest.dtype == np.uint16:
+            dest = dest.astype(np.int32)
+        elif dest.dtype == np.uint32:
+            dest = dest.astype(np.int64)
+
+        tensor = torch.tensor(dest)
+
+        return tensor
+
+
+class GDLVectorDataset(GeoDataset):
+    """Abstract base class for :class:`GeoDataset` stored as vector files."""
+    def __init__(self, vec_ds: str = None, res: float = 0.0001) -> None:
+        """Initialize a new Dataset instance.
+
+        Args:
+            vec_ds: vector labels geopackage
+            res: resolution of the dataset in units of CRS
+        Returns:
+            An OGR datasource in memory
+        """
+        super().__init__()
+
+        self.vec_ds = ogr.Open(str(vec_ds))
+        self.res = res
+
+        # Populate the dataset index
+        try:
+            with fiona.open(str(vec_ds)) as src:
+                minx, miny, maxx, maxy = src.bounds
+
+        except fiona.errors.FionaValueError:
+            print('Vector file could not be read.')
+
+        mint = 0
+        maxt = sys.maxsize
+        coords = (minx, maxx, miny, maxy, mint, maxt)
+        self.index.insert(0, coords, 'vec')
+
+        src_mem_driver = ogr.GetDriverByName('MEMORY')
+        self.mem_vec_ds = src_mem_driver.CopyDataSource(self.vec_ds, 'src_mem_ds')
+        self.vec_srs = self.mem_vec_ds.GetLayer().GetSpatialRef()
+        vec_srs_wkt = self.vec_srs.ExportToPrettyWkt()
+
+        self._crs = CRS.from_wkt(vec_srs_wkt)
+
+    def __getitem__(self, query: BoundingBox) -> Dict[str, Any]:
+        """Retrieve image/mask and metadata indexed by query.
+
+        Args:
+            query: (minx, maxx, miny, maxy, mint, maxt) coordinates to index
+
+        Returns:
+            sample as a OGR datasource in memory and metadata at that index
+        """
+        poly_box = ogr.Geometry(ogr.wkbLinearRing)
+        poly_box.AddPoint(query.minx, query.maxy)
+        poly_box.AddPoint(query.maxx, query.maxy)
+        poly_box.AddPoint(query.maxx, query.miny)
+        poly_box.AddPoint(query.minx, query.miny)
+        poly_box.AddPoint(query.minx, query.maxy)
+        # Create a Polygon object from the ring.
+        poly = ogr.Geometry(ogr.wkbPolygon)
+        poly.AddGeometry(poly_box)
+
+        # # Create a vector datasource in memory:
+        mem_driver = ogr.GetDriverByName('MEMORY')
+        mem_ds = mem_driver.CreateDataSource('memdata')
+        mem_layer = mem_ds.CreateLayer('0', self.vec_srs, geom_type=ogr.wkbPolygon)
```

Comment:
I'm not very familiar with Ogr. It's cool to see this implementation. However, as commented above, could we consider using geopandas to make uniform the librairies we use when manipulating vector data? I don't think it's necessary to switch to geopandas in this PR however. We can merge and have this discussion afterwards if we think it's relevant.

Reply:
Could not implement it with geopandas because it did not properly transferred the scheme of the initial vector file to the patches. Also, using the OGR  we can directly manipulate in-memory datasets which is 1. explicit, 2. explainable, and 3. fast.